nistats/first_level_model.py

```diff
@@ -467,23 +479,33 @@ def compute_contrast(self, contrast_def, contrast_name=None,
 
         Returns
         -------
-        output_image : Nifti1Image
-            The desired output image
+        output : Nifti1Image
+            The desired output image from a fixed effects computation.
 
         """
         if self.labels_ is None or self.results_ is None:
             raise ValueError('The model has not been fit yet')
 
-        if isinstance(contrast_def, np.ndarray):
+        if isinstance(contrast_def, (np.ndarray, str)):
             con_vals = [contrast_def]
         elif isinstance(contrast_def, (list, tuple)):
             con_vals = contrast_def
-            for cidx, con in enumerate(contrast_def):
+        else:
+            raise ValueError('contrast_def must be an array or str or list of'
+                             ' (array or str)')
+        # We reinterpret formulas as vectors only if design has column names.
+        des_names = np.unique(self.design_matrices_[0].columns.tolist())
+        if len(des_names) == len(self.design_matrices_[0].columns):
+            di = DesignInfo(self.design_matrices_[0].columns.tolist())
+            for cidx, con in enumerate(con_vals):
                 if not isinstance(con, np.ndarray):
-                    raise ValueError('contrast_def at index %i is not an'
-                                     ' array' % cidx)
+                    con_vals[cidx] = di.linear_constraint(con).coefs
         else:
-            raise ValueError('contrast_def must be an array or list of arrays')
+            for cidx, con in enumerate(con_vals):
+                if not isinstance(con, np.ndarray):
+                    raise ValueError('can not interpret formula when design'
+                                     ' do not contain columns with unique '
+                                     'names')
```

Comment:
Hm, thinking a bit further: we certainly don't want design to contain columns with redundant identifiers, so this should be enforced in the design matrix construction, not here (I find this code is becoming too convoluted because we are taking care of such issues that should be handled elsewhere)